### PR TITLE
Selection of dependent monitors

### LINKED ISSUE #1089: create parent and child for notification
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

Other

### 🔖 Feature description

Hello there,

may it's possible to create a parent (internet connection) and children (server behind this router).
I always get a bunch of service notifications when the internet is down. 

### ✔️ Solution

I would like to create a parent group (Starlink and AT&T e.g.), so when the internet connection should be down, I only get one message for the internet Service and not for my 20 servers running, which aren't reachable due to the internet connection.

### ❓ Alternatives

_No response_

### 📝 Additional Context

Anyway, great tool, Thanks a lot!

### LINKED ISSUE #1534: Bundle notifications
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Notification

### 🔖 Feature description

W're using the email notification and just received over 100 mails at a time when a server went down.
It would be nice if the notifications could be bundled to a single email.
This could be beneficial for other notification methods as well.

### ✔️ Solution

In my use case, you could wait 10 sec (or less, maybe make it a setting) before sending an e-mail and check if other notifications are being fired as well.

### ❓ Alternatives

I have considered other notification methods, but we only ever use e-mail

### 📝 Additional Context

We have around 400 websites setup 😅

### LINKED ISSUE #3238: [groups] Monitor critical flag to replace 'Partially Degraded Service' with 'Full Service Outage'
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

The Status Page on top shows `All Systems Operational` or `Partially Degraded Service` but there is never really a `Degraded Service` or `Full Service Outage` message, since that requires all Monitors to be down ?

I have a bunch of Monitors and some like `Router` or `Firewall` can be interpreted as if they are down nothing works. There are still other Monitors like redundancy or offsite services that will always be up though.

I think it would be great if we had a `Critical Service` checkbox (or dropdown with levels?) on the Monitor.

### ✔️ Solution

Adding a checkbox `Critical Service` on Monitors.

If this is checked and the Monitor is on the Status Page with Status `down` the heading message changes to `Full Service Outage` with red error symbol.

Actual message e.g. `Full Service Outage` to be defined.

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by karelkryda**
# Description

Added the ability to choose which monitors the current monitor depends on.
For example, if we have website monitoring and its functionality depends on the database server, in the event of a database server failure, the status of the website is automatically set to "DEGRADED".

TODO:
- [x] Automatic detection of the status of the slave monitor, according to the status of the master monitor
- [x] Optionally (do not) send notifications if the master monitor is PENDING, DOWN or DEGRADED and the slave monitor is also DOWN (and also if it was DOWN - DEGRADED and now it is UP)

## Type of change

Please delete any options that are not relevant.

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

image
image

Resolves #1089
Resolves #1534
Resolves #3238

### CODE REVIEW

**Comment by bicisteadm** [comment]
When will this feature be added? It's amazing and I definitely must have 👍

**Comment by Saibamen** [comment]
@louislam

**Comment by mathiskir** [comment]
Hey,
I just tried using this feature, but it seems not to work.
I installed and started it with `git clone -b dependent-monitors https://github.com/karelkryda/uptime-kuma && cd uptime-kuma && npm ci && npm run build && node server/server.js`
And I got this error after I added a test monitor:
image
I just wanted to set up a blank monitor without any dependency

**Comment by mathiskir** [comment]
Works now.

**Comment by stefou84** [comment]
@louislam  pleaseee :p

**Comment by SAOPP** [comment]
Guys I need it! :)

**Comment by acki** [comment]
@louislam Please ;-)

**Comment by gaby** [comment]
@louislam Thoughts on this? it's been open for 3 months.

**Comment by karelkryda** [comment]
@louislam, Should I prepare this PR for merge (resolve conflicts) or do you not plan to merge this PR now?

File: server/model/monitor.js [hunk 1]
```diff
@@ -450,6 +463,11 @@ class Monitor extends BeanModel {
 
                 retries = 0;
 
+                if (bean.status === UP && isDegraded) {
+                    bean.msg = "Monitor is degraded, because at least one master monitor is pending, down or degraded";
+                    bean.status = DEGRADED;
+                }
+
```
Comment (Saibamen) [inline server/model/monitor.js:470]:
```suggestion
                }
```

File: server/model/monitor.js [hunk 2]
```diff
@@ -463,6 +481,12 @@ class Monitor extends BeanModel {
                     retries++;
                     bean.status = PENDING;
                 }
+
+                // Do not change this text!
+                // Condition below and in api-router depends on it
```
Comment (Saibamen) [inline server/model/monitor.js:486]: Can you use some global constant for this text in `monitor.js` and `api-router.js`

**Review by Saibamen — CHANGES_REQUESTED** [review]

File: server/model/monitor.js (hunk 2 above)
Comment (karelkryda) [reply server/model/monitor.js]: Added `MONITOR_DOWN_DEGRADED` variable to utils.ts/.js

**Comment by Saibamen** [comment]
I didn't test it, but it is possible to have two monitors with are depended on each other?

For example, monitor x and y.
x have set dependent monitor as y and y has dependent monitor set to x.

IMO this should not be possible

**Comment by karelkryda** [comment]
> I didn't test it, but it is possible to have two monitors with are depended on each other?
> 
> For example, monitor x and y. x have set dependent monitor as y and y has dependent monitor set to x.
> 
> IMO this should not be possible

In fact, I think this is possible. I wonder if this is a problem or not, or how to solve it.

**Comment by Saibamen** [comment]
Just simply check (in `addDependentMonitors` -> change name to `checkAndAddDependentMonitors` maybe?) if requested dependent monitors
```
monitor_id: monitorID,
depends_on: monitor.id,
```
(`depends_on`) already has set its own `depends_on` to `monitorID`?

File: src/pages/EditMonitor.vue [hunk 3]
```diff
@@ -475,6 +512,17 @@ export default {
...
+        this.$root.getMonitorList((res) => {
+            if (res.ok) {
+                Object.values(this.$root.monitorList).filter(monitor => monitor.id !== Number(this.$route.params.id)).map(monitor => {
+                    this.dependentMonitorsOptions.push({
+                        id: monitor.id,
+                        name: monitor.name,
+                    });
+                });
+            }
+        });
```
Comment (karelkryda) [inline src/pages/EditMonitor.vue:524]: @Saibamen I think I will adjust this function to load all monitors from the database, except those that depend on that monitor.

What do you think about it?

Comment (karelkryda) [reply src/pages/EditMonitor.vue:524]: This will prevent monitor y from relying on monitor x if monitor x is already dependent on monitor y.

Comment (Saibamen) [reply src/pages/EditMonitor.vue:524]: Looks good

Comment (karelkryda) [reply src/pages/EditMonitor.vue]: @Saibamen should be OK now 🙂

**Review by Saibamen — APPROVED** [review]

**Comment by karelkryda** [comment]
@louislam Do you think this is okay for milestone 1.16.0 or 1.17.0?

**Comment by jeremyrnelson** [comment]
@karelkryda, this feature looks amazing!  May I make a few suggestions?

1. I find your use of the word "master" confusing.  Can I suggest maybe replace this wording: "Select the monitor(s) on which this monitor depends. If the master monitor(s) fails, this monitor will be affected too." with "Select the monitor(s) on which this monitor depends - if any dependency monitor fails, this monitor will be marked DEGRADED."

Also "Monitor is degraded, because at least one master monitor is pending, down or degraded" with "Monitor is degraded, because at least one dependency monitor is pending, down or degraded"

2. Can you create a new "Virtual" (or maybe even Dummy or Master) check type that doesn't do anything itself, but just effectively reports on the status of dependencies?  In other words, UP means all dependency monitors are UP, DEGRADED means 1 or more dependency monitors is down, and DOWN means all dependency monitors are DOWN.  Even just pinging localhost wouldn't really accomplish this goal, since it would only show DEGRADED even if all the dependency monitors are DOWN.

This would be super helpful for a couple of scenarios:

- Say a host "webserver" quits answering pings, but is still responding to http requests.  I want a "virtual" monitor that says that this host is degraded.
- Say I'm monitoring 3 physical locations (A, B, and C) with 3 hosts in each (A1, A2, A3, B1, B2, B3, C1, C2, C3).  If location A is completely offline, end users won't understand or care about A1, A2, and A3 being offline - they just want to see a check that say "Location A" DOWN.  There's no specific check that could tell if A is down - a user doesn't want to see a barrage of messages about various hosts being up or down, they want to see the status of that location.

**Comment by stelle007** [comment]
Did this ever get added?

**Comment by Furrypaw** [comment]
I have not tried this but would following be possible:
A - Network
B - Service
C - Service

So lets say A goes down, currently I would get notification for A; B; C are being down. Now it would be nice if with this I could get 1 message saying that A is down and B;C are being affected.

**Comment by mabed-fr** [comment]
@louislam Can the functionality be integrated ?

**Comment by mabed-fr** [comment]
@louislam Is this feature soon to be integrated?

**Comment by SAOPP** [comment]
Interesting too.

**Comment by mabed-fr** [comment]
This MR julian-piehl:group-monitors not does not concretely meet the parent-child need (hierarchy)

**Comment by CommanderStorm** [comment]
@karelkryda 
Could you add The following to the description (allowing these to show up as having a PR ready to merge them ^^):
Resolves #1089
Resolves #2261
Resolves #2348
Resolves #1887
Resolves #1534
Resolves #3238
Resolves #2335

(there are likely more that can be linked to the description, but these were the ones I found at a cursory glance)

**Comment by CommanderStorm** [comment]
Yea, I missed:
Resolves #2487

**Comment by elboletaire** [comment]
Any news on this? It's a really interesting functionality and I'd love to be using it already 😇

**Comment by DANW999** [comment]
Any update on this at all? This would be a great feature. Currently, I have a few services behind my reverse proxy and therefore get alerts for all of them when the reverse proxy goes down. It would be ideal if I could make all services a dependency of the reverse proxy such that when the proxy goes down, I only get an alert for the proxy itself rather than everything else behind it as well.

**Comment by iawa2k** [comment]
@karelkryda Any plans to make this compatible with the latest release so it can be merged in? Really useful tool and the one thing that is missing and keeping me from fully switching from Woodstone’s Server Alive 😅

**Comment by karelkryda** [comment]
> @karelkryda Any plans to make this compatible with the latest release so it can be merged in? Really useful tool and the one thing that is missing and keeping me from fully switching from Woodstone’s Server Alive 😅

Truth be told, I've always been waiting for the moment when Luis would be ready to merge my PRs in a way that would prevent them from becoming obsolete again (because they won't be merged).

Since I don't see interest in merging, it doesn't make sense for me to edit anything at the moment. I don't have a problem with Luis editing and merging his PR, but I don't have enough free time to do it at the moment.

**Comment by mh166** [comment]
@louislam Since you are planning v2.0 at the moment, would it be possible for you to consider this PR? It will allow for an even more fine-grained monitoring and, more importantly, notification setup. Right now I get notifications for _"down"_ alerts that really should be _"degraded"_ informations.

**Two examples:**
1. We check two target hosts in order to make sure each of our internet connections is available. If either `main` or `backup` are failing, the whole `Internet connection` group is failing – even though just one of our connections dropped and the other might still serve as a failover.
2. Personally, in my home I have a number of smart home devices that I monitor. If one of them is unreachable, the whole `Smart Home` group is shown as _"down"_, even though the rest of the devices is still working fine.

It would be very much appreciated if you could give a short notice on wether you're considering this PR for v2.0 (or even some minor version shortly thereafter). I'm sure, @karelkryda will be happy to rebase his PR if there is a merge to be expected.

Thank you so much for your consideration and, even more so, for this amazing tool! 🤩

**Comment by CommanderStorm** [comment]
> Since you are planning v2.0 at the moment

v2.0 planning is final and adding new big features at this point is not possible.
We are currently concentrating our efforts on fixing remaining bugs and are working towards the first beta release.
See #4171 for further details on the still nessesary bugfixes

> wether you're considering this PR for v2.0 (or even some minor version shortly thereafter)

Currently, I think this PR not only needs a rebasing, but also a reworking of the proposed functionality.

Concretely, this functionality should be integrated into the group monitor.
I think this should also include making the group monitor into a notification-provider according to our modular (and more importantly maintainable) syntax: https://github.com/louislam/uptime-kuma/tree/master/server/monitor-types
(ideally in a different PR for reviewability if possible). 
Refactoring the group monitor is something that we could be integrated into the [`v2.1` merge slot](https://github.com/louislam/uptime-kuma/milestone/40)

For this feature concretely, I think the [`v2.2` merge slot](https://github.com/louislam/uptime-kuma/milestone/42) could be possible, as in [`v2.1` we will concentrate on smaller bugfixes, notification providers and the api](https://github.com/louislam/uptime-kuma/milestone/40) and thus already have a lot of big features in there.

**Comment by CommanderStorm** [comment]
I am going to close this issue as I don't see additional value in keeping this open.
Currently, I don't see the conflicts in this PR being resolved and this PR being merged

Making the group monitor into a notification-provider according to [our modular (and more importantly maintainable) syntax](https://github.com/louislam/uptime-kuma/tree/master/server/monitor-types) is included in the `v2.1` merge slot.
The PR in question is #4395

Further work is still needed, and help in this area (based on #4395) would still be appreciated.
Our [contribution guide can be found here](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md)

**Comment by karelkryda** [comment]
@CommanderStorm To be honest, I'm not entirely sure I understand the reason for closing this PR. While I'm not familiar with the changes in version 2.0+, I couldn't find anything in the PRs for version 2.1 that would explain the closure of this PR.
Could you please explain to me the reason for the closure of this PR? 

Thank you

**Comment by CommanderStorm** [comment]
#4395 redactors the code into a single module
=> making working/maintainance in this area much easier.

**Comment by wsw70** [comment]
> Currently, I have a few services behind my reverse proxy and therefore get alerts for all of them when the reverse proxy goes down. It would be ideal if I could make all services a dependency of the reverse proxy such that when the proxy goes down, I only get an alert for the proxy itself rather than everything else behind it as well.

I have the exact same situation (and would rely on a ping (ICMP) to check if the line is up to start with). I just discovered uptime-kuma and before diving down I would be glad to check if the dependency feature is available (I did not find anything in the docs)

**Comment by CommanderStorm** [comment]
> I would be glad to check if the dependency feature is available (I did not find anything in the docs)

@wsw70 please see my comment above: https://github.com/louislam/uptime-kuma/pull/1236#issuecomment-1924740307
TLDR: no, we currently do not offer such a feature
